src/Functions/array/arrayIntersect.cpp

```diff
@@ -549,44 +550,101 @@ ColumnPtr FunctionArrayIntersect::execute(const UnpackedArrays & arrays, Mutable
+        if (arg.is_const)
+            off = (*arg.offsets)[0];
+        else
+            off = (*arg.offsets)[row];
 
-        for (const auto & pair : map)
+        for (auto i : collections::range(prev_off[0], off))
         {
-            if (pair.getMapped() == args)
+            all_has_nullable = all_nullable;
+            current_has_nullable = false;
+            typename Map::LookupResult pair = nullptr;
+
+            if (arg.null_map && (*arg.null_map)[i])
+            {
+                current_has_nullable = true;
+                if (null_added)
+                    continue;
+            }
+            else if constexpr (is_numeric_column)
+            {
+                pair = map.find(columns[0]->getElement(i));
+            }
+            else if constexpr (std::is_same_v<ColumnType, ColumnString> || std::is_same_v<ColumnType, ColumnFixedString>)
+                pair = map.find(columns[0]->getDataAt(i));
+            else
+            {
+                const char * data = nullptr;
+                pair = map.find(columns[0]->serializeValueIntoArena(i, arena, data));
+            }
+            prev_off[0] = off;
+            if (arg.is_const)
+                prev_off[0] = 0;
+
+            if (!current_has_nullable)
+                all_has_nullable = false;
+
+            if (pair && pair->getMapped() == args)
             {
+                ++pair->getMapped();
                 ++result_offset;
                 if constexpr (is_numeric_column)
-                    result_data.insertValue(pair.getKey());
+                {
+                    if (pair->getKey() == columns[0]->getElement(i))
+                    {
+                        result_data.insertValue(pair->getKey());
+                    }
+                }
                 else if constexpr (std::is_same_v<ColumnType, ColumnString> || std::is_same_v<ColumnType, ColumnFixedString>)
-                    result_data.insertData(pair.getKey().data, pair.getKey().size);
+                {
+                    if (pair->getKey() == columns[0]->getDataAt(i))
+                    {
+                        result_data.insertData(pair->getKey().data, pair->getKey().size);
+                    }
+                }
                 else
-                    result_data.deserializeAndInsertFromArena(pair.getKey().data);
-
+                {
+                    const char * data = nullptr;
+                    if (pair->getKey() == columns[0]->serializeValueIntoArena(i, arena, data))
+                    {
+                        result_data.deserializeAndInsertFromArena(pair->getKey().data);
+                    }
+                }
                 if (all_nullable)
                     null_map.push_back(0);
             }
+            if (all_has_nullable && !null_added)
+            {
+                ++result_offset;
+                result_data.insertDefault();
+                null_map.push_back(1);
+                null_added = true;
+            }
```

Comment:
This is not very important, but I would prefer to move this block of code earlier. E.g. at line 581.